src/tests/system.rs

```diff
@@ -607,3 +576,195 @@ mod fixed {
         }
     }
 }
+
+// Somehow Complex<f32|f64> don't implement these methods, even though they
+// should, bc f32 / f64 implement Float and FloatCore:
+//  - https://docs.rs/num/latest/num/struct.Complex.html#impl-5
+//  - https://docs.rs/num/latest/num/struct.Complex.html#impl-4
+//
+// See discussion here:
+// https://github.com/iliekturtles/uom/pull/287#pullrequestreview-922005933
+/*
+mod complex {
```

Comment:
I spent some time looking at this tonight and I believe what we'll need to do is explicitly implement these functions for `f32`, `f64`, `Complex32`, and `Complex64` using the `storage_types!` macro. `Complex<T>` doesn't implement `Float`/`FloatCore`, but it does implement these methods when the underlying type, `T`, implements `FloatCore`.

```rust
storage_types! {
    types: Float, Complex;

    impl<...> Quantity<..., V /*V type alias defined by storage_types!*/> {
        fn is_infinite(self) -> bool { self.value.is_infinite() }
    }
}

// I don't think a separate storage_types! is necessary. The code should be able to be the same.
//storage_types! {
//    types: Complex;
//    
//    impl<...> Quantity<..., V> {
//        fn is_infinite(...) -> ... { ... }
//    }
}

Reply:
I don't completely understand why I should implement these methods on `f32` and `f64` too, num implements them for these two just fine, doesn't it?

As for doing it for the complexes... it's weird. Before the compiler was complaining that it didn't find an implementation for `is_infinite` et al., but now when I implement it with the following patch, it tells me it has duplicate implementation:

```diff
diff --git a/src/tests/system.rs b/src/tests/system.rs
index 1355f80..cad6de9 100644
--- a/src/tests/system.rs
+++ b/src/tests/system.rs
@@ -584,7 +584,6 @@ mod fixed {
 //
 // See discussion here:
 // https://github.com/iliekturtles/uom/pull/287#pullrequestreview-922005933
-/*
 mod complex {
     storage_types! {
         types: Complex;
@@ -593,6 +592,11 @@ mod complex {
 
         Q!(crate::tests, V);
 
+        impl<D, U> Quantity<D, U, V /*V type alias defined by storage_types!*/> {
+            fn is_infinite(self) -> bool { self.value.is_infinite() }
+            fn is_finite(self) -> bool { self.value.is_finite() }
+        }
+
         #[test]
         fn fp_categories() {
             let float_categories = vec![
@@ -724,7 +728,6 @@ mod complex {
         }
     }
 }
-*/
 
 mod non_complex {
     storage_types! {

```

```
error[E0592]: duplicate definitions with name `is_infinite`
    |
   ::: src/tests/system.rs:596:13
    |
596 |               fn is_infinite(self) -> bool { self.value.is_infinite() }
    |               ---------------------------- other definition for `is_infinite`
    |
   ::: src/tests/mod.rs:55:1
    |
55  | / system! {
56  | |     quantities: Q {
57  | |         length: meter, L;
58  | |         mass: kilogram, M;
...   |
65  | |     }
66  | | }
    | |_- in this macro invocation
   --> src/system.rs:613:13
    |
613 | /             pub fn is_infinite(self) -> bool
614 | |             where
615 | |                 V: $crate::num::Float,
    | |______________________________________^ duplicate definitions for `is_infinite`
    |
    = note: this error originates in the macro `system` (in Nightly builds, run with -Z macro-backtrace for more info)
```

Maybe I am also implementing it at the wrong location?

Reply:
I think I see where the duplicates are and will take a look in the next few days!